### PR TITLE
Fixes incorrect read/write permissions on Flutter.framework and FlutterMacOS.framework

### LINKED ISSUE #147142: Building Flutter projects on MacOS fails if FlutterMacOS.framwork is read-only (ie. in a Nix environment)
### Steps to reproduce

1. Make sure `nix` is installed on your MacOS machine
2. Run `nix-shell -p flutter` to create a shell with Flutter in your environment
3. Attempt to build a project for MacOS with `flutter build macos` and notice how it fails
4. Leave the `nix` environment you created in step 2 (`exit`)
5. Attempt to build again for MacOS with `flutter build macos` and notice how it succeeds

### Expected results

The build should succeed under the nix environment.

### Actual results

The build fails on the nix environment due to `FlutterMacOS.framework` being read-only in the Nix store. The Nix store is intentionally read-only, so changing `FlutterMacOS.framework` to `rw` within the Nix store would ruin the point of using Nix here.

### Code sample

<details open><summary>Code sample</summary>
As this issue occurs on the brand-new project created by `dart create bug`, I attached that file unmodified:

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a purple toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

</details>

### Screenshots or Video

_No response_

### Logs

<details open><summary>Logs</summary>

```console

underworld:bug pearl$ flutter build macos

2024-04-21 16:23:38.484 xcodebuild[49965:156303515] DVTCoreDeviceEnabledState: DVTCoreDeviceEnabledState_Disabled set via user default (DVTEnableCoreDevice=disabled)
--- xcodebuild: WARNING: Using the first of multiple matching destinations:
{ platform:macOS, arch:arm64, id:00006000-001230823E7A401E }
{ platform:macOS, arch:x86_64, id:00006000-001230823E7A401E }
Building App.framework for x86_64...
Building App.framework for arm64...
Target release_unpack_macos failed: Exception: Failed to extract arm64 x86_64 for /Users/pearl/code/test/bug/build/macos/Build/Products/Release/FlutterMacOS.framework/Versions/A/FlutterMacOS.
fatal error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/lipo: can't create temporary output file: /Users/pearl/code/test/bug/build/macos/Build/Products/Release/FlutterMacOS.framework/Versions/A/FlutterMacOS.lipo (Permission denied)

Running lipo -info:
Architectures in the fat file: /Users/pearl/code/test/bug/build/macos/Build/Products/Release/FlutterMacOS.framework/Versions/A/FlutterMacOS are: x86_64 arm64 


Command PhaseScriptExecution failed with a nonzero exit code
warning: Run script build phase 'Run Script' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase. (in target 'Flutter Assemble' from project 'Runner')
** BUILD FAILED **

Building macOS application...                                           
Build process failed
underworld:bug pearl$ 
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
underworld:bug pearl$ flutter doctor -v
[✓] Flutter (Channel stable, 3.19.0, on macOS 13.4.1 22F770820d darwin-arm64, locale en-US)
    • Flutter version 3.19.0 on channel stable at /nix/store/yj4imxhd3lcv6gjvv584ggchqy9k4krx-flutter-wrapped-3.19.0-sdk-links
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision nixpkgs000 (), 1970-01-01 00:00:00
    • Engine revision 04817c99c9
    • Dart version 3.3.0
    • DevTools version 2.31.1

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /nix/store/z0mkx48izmj3iv10p3ig4ij2qd3iq1j7-androidsdk/libexec/android-sdk
    • Platform android-34, build-tools 30.0.3
    • ANDROID_HOME = /nix/store/z0mkx48izmj3iv10p3ig4ij2qd3iq1j7-androidsdk/libexec/android-sdk
    • Java binary at: /Users/pearl/Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)
    ! Some Android licenses not accepted. To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.15.2

[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 2023.2)
    • Android Studio at /Users/pearl/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0-17.0.9b1087.7-11185874)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.3.4)
    • IntelliJ at /Users/pearl/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 78.4.2
    • Dart plugin version 233.15123

[✓] IntelliJ IDEA Community Edition (version 2024.1)
    • IntelliJ at /Users/pearl/94w87dhc6smqxdlq61kvy2ksd99j1fhw-idea-community-2024.1/Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2024.1)
    • IntelliJ at /Users/pearl/z9dfy1vs6g4gybvznyqlxn6453mgsy4z-idea-ultimate-2024.1/Applications/IntelliJ IDEA.app
    • Flutter plugin version 79.0.3
    • Dart plugin version 241.15845

[✓] VS Code (version 1.88.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (1 available)
    • macOS (desktop) • macos • darwin-arm64 • macOS 13.4.1 22F770820d darwin-arm64

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
```

</details>

### LINKED ISSUE #148354: MacOS: Codemagic - The installer package includes files that are only readable by the root user (3.22.0)
### Steps to reproduce

Regression: https://github.com/flutter/flutter/issues/132725

Follow https://docs.flutter.dev/deployment/macos#create-a-build-archive-with-codemagic-cli-tools up to and including app-store-connect publish --path "$PACKAGE_NAME"

### Expected results

Publishing to the App Store works

### Actual results

2024-05-14 21:01:55.328 *** Error: ERROR: [ContentDelivery.Uploader] Asset validation failed (90255) The installer package includes files that are only readable by the root user. This will prevent verification of the application's code signature when your app is run. Ensure that non-root users can read the files in your app.

### Code sample

I haven't been able to reproduce this since to get to the point of publishing to the App Store, you need to set up all of the certificates etc. I'm guessing this occurs in a blank Flutter project though.

I can confirm, however, that reverting to Flutter 3.19.6 doesn't have this issue.

Also, I've tested on master channel - it does not work too.

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

```
== Use Xcode 15.3 ==
> sudo xcode-select --switch /Applications/Xcode-15.3.app/Contents/Developer
Xcode 15.3
Build version 15E204a

== Use default version of CocoaPods ==
> gem list '^cocoapods$' --no-verbose
cocoapods (1.14.2)

== Use Java version 11 ==
> java -version
openjdk version "11.0.21" 2023-10-17 LTS
OpenJDK Runtime Environment Zulu11.68+17-CA (build 11.0.21+9-LTS)
OpenJDK 64-Bit Server VM Zulu11.68+17-CA (build 11.0.21+9-LTS, mixed mode)

== Codemagic CLI tools ==
> codemagic-cli-tools version
codemagic-cli-tools 0.51.0
```

```
== Use Flutter channel stable ==
> curl --fail --show-error --silent --output flutter_macos_arm64_3.19.6-stable.zip --connect-timeout 10.00 --max-time 120.00 https://storage.googleapis.com/flutter_infra_release/releases/stable/macos/flutter_macos_arm64_3.19.6-stable.zip

> 7z -bd x flutter_macos_arm64_3.19.6-stable.zip

7-Zip [64] 17.04 : Copyright (c) 1999-2021 Igor Pavlov : 2017-08-28
p7zip Version 17.04 (locale=utf8,Utf16=on,HugeFiles=on,64 bits,4 CPUs LE)

Scanning the drive for archives:
1 file, 1169959256 bytes (1116 MiB)

Extracting archive: flutter_macos_arm64_3.19.6-stable.zip
--
Path = flutter_macos_arm64_3.19.6-stable.zip
Type = zip
Physical Size = 1169959256

Everything is Ok

Folders: 2795
Files: 10321
Size:       2650462486
Compressed: 1169959256

Flutter 3.19.6 • Dart 3.3.4
```

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console

[✓] Flutter (Channel master, 3.22.0-33.0.pre.20, on macOS 14.3.1 23D60 darwin-arm64, locale en-US)
    • Flutter version 3.22.0-33.0.pre.20 on channel master at /Users/builder/programs/flutter_3_16_7
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 4a1e3eaaa2 (26 minutes ago), 2024-05-14 15:58:18 -0500
    • Engine revision ae9ff69a08
    • Dart version 3.5.0 (build 3.5.0-153.0.dev)
    • DevTools version 2.36.0-dev.5

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /usr/local/share/android-sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /usr/local/share/android-sdk
    • ANDROID_SDK_ROOT = /usr/local/share/android-sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.3)
    • Xcode at /Applications/Xcode-15.3.app/Contents/Developer
    • Build 15E204a
    • CocoaPods version 1.14.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.3.1 23D60 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.3.1 23D60 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 123.0.6312.58

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by LouiseHsu**
Fixes https://github.com/flutter/flutter/issues/148354

Fixes https://github.com/flutter/flutter/issues/147142
Closes https://github.com/flutter/flutter/pull/147144

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by nilsreichardt** [comment]
Thanks for the fix! When this fix is merged, can we cherry pick it and release it as a hotfix since many Flutter users are affected and can't publish MacOS apps with Flutter v3.22.0 anymore?

* https://github.com/flutter/flutter/issues/147142
* https://github.com/flutter/flutter/issues/148354

File: packages/flutter_tools/test/host_cross_arch.shard/macos_content_validation_test.dart [hunk 1]
```diff
@@ -164,6 +164,20 @@ void main() {
...
       );
 
+      final File flutterMacOSFramework = fileSystem.file(
+        fileSystem.path.join(
+          outputApp.path,
+          'Contents',
+          'Frameworks',
+          'FlutterMacOS.framework',
+        ),
+      );
```
Comment (jmagman) [inline packages/flutter_tools/test/host_cross_arch.shard/macos_content_validation_test.dart:174]: `outputFlutterFramework` is the same thing immediately above, so you should be able to use that.  And it's a directory, not a file.

File: packages/flutter_tools/test/host_cross_arch.shard/ios_content_validation_test.dart [hunk 2]
```diff
@@ -456,6 +456,19 @@ void main() {
         ],
       );
       expect(appCodesign, const ProcessResultMatcher());
+
+      final File flutterIOSFramework = fileSystem.file(
```
Comment (jmagman) [inline packages/flutter_tools/test/host_cross_arch.shard/ios_content_validation_test.dart:460]: `flutterFramework` is already the same thing, you should be able to use that, and it's a directory.

**Review by jmagman — APPROVED** [review]
LGTM

File: packages/flutter_tools/test/host_cross_arch.shard/ios_content_validation_test.dart [hunk 3]
```diff
@@ -422,10 +422,17 @@ void main() {
...
+        ),
+      );
+
       final String flutterFramework = fileSystem.path.join(
         appBundle.path,
-        'Frameworks',
-        'Flutter.framework',
+        flutterFrameworkDir.path,
         'Flutter',
       );
```
Comment (jmagman) [inline packages/flutter_tools/test/host_cross_arch.shard/ios_content_validation_test.dart:437]: Shouldn't this be:
```
      final String flutterFramework = fileSystem.path.join(
        flutterFrameworkDir.path,
        'Flutter',
      );
```

**Comment by jmagman** [comment]
> Thanks for the fix! When this fix is merged, can we cherry pick it and release it as a hotfix since many Flutter users are affected and can't publish MacOS apps with Flutter v3.22.0 anymore?

@nilsreichardt Yes, this will be hotfixed to 3.22.  More details at https://github.com/flutter/flutter/issues/148354#issuecomment-2122982807

**Comment by nilsreichardt** [comment]
> @nilsreichardt Yes, this will be hotfixed to 3.22. More details at [#148354 (comment)](https://github.com/flutter/flutter/issues/148354#issuecomment-2122982807)

Thanks, that's awesome <3

**Comment by jmagman** [comment]
@nilsreichardt Thanks for validating!  You're right about the changelog, I'll follow up.

**Pull request merged**